// @keystone-next/keystone/___internal-do-not-use-will-break-in-patch/next-graphql/dist/keystone.esm.js
import { i as initConfig, c as createSystem } from '../../../dist/initConfig-dae1a62a.esm.js';
import { c as createApolloServerMicro } from '../../../dist/createApolloServer-c5812228.esm.js';
import '@babel/runtime/helpers/objectSpread2';
import '../../../dist/core-3f0f7b15.esm.js';
import '../../../dist/sqlite-335dc173.esm.js';
import '../../../dist/graphql-ts-schema-e8d59289.esm.js';
import '@graphql-ts/schema';
import 'graphql-type-json';
import 'graphql-upload/public/GraphQLUpload.js';
import 'graphql';
import 'decimal.js';
import '@graphql-ts/schema/api-without-context';
import '@graphql-ts/schema/api-with-context';
import '@graphql-tools/schema';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../../dist/graphql-errors-b69877ed.esm.js';
import 'apollo-server-errors';
import 'path';
import 'fs-extra';
import 'fast-glob';
import 'prettier';
import 'resolve';
import '@emotion/hash';
import '../../../dist/admin-meta-graphql-433469b3.esm.js';
import '@apollo/client';
import '../../../session/dist/keystone.esm.js';
import 'cookie';
import '@hapi/iron';
import 'uid-safe';
import '@babel/runtime/helpers/classPrivateFieldInitSpec';
import '@babel/runtime/helpers/classPrivateFieldGet';
import '@babel/runtime/helpers/classPrivateFieldSet';
import 'p-limit';
import '../../../dist/types-for-lists-20b6587f.esm.js';
import 'graphql/execution/values';
import 'pluralize';
import 'uuid';
import 'image-type';
import 'image-size';
import '../../../fields/types/image/utils/dist/keystone.esm.js';
import 'node-fetch';
import 'form-data';
import 'crypto';
import 'stream';
import 'filenamify';
import '@sindresorhus/slugify';
import '../../../fields/types/file/utils/dist/keystone.esm.js';
import 'cuid';
import '../../../dist/package-path-7ba957ae.esm.js';
import 'apollo-server-micro';
import 'apollo-server-express';

function nextGraphQLAPIRoute(keystoneConfig, prismaClient) {
  var _keystoneConfig$graph;

  const initializedKeystoneConfig = initConfig(keystoneConfig);
  const {
    graphQLSchema,
    getKeystone
  } = createSystem(initializedKeystoneConfig);
  const keystone = getKeystone(prismaClient);
  keystone.connect();
  const apolloServer = createApolloServerMicro({
    graphQLSchema,
    createContext: keystone.createContext,
    sessionStrategy: initializedKeystoneConfig.session,
    graphqlConfig: initializedKeystoneConfig.graphql,
    connectionPromise: keystone.connect()
  });
  return apolloServer.createHandler({
    path: ((_keystoneConfig$graph = keystoneConfig.graphql) === null || _keystoneConfig$graph === void 0 ? void 0 : _keystoneConfig$graph.path) || '/api/graphql'
  });
}

export { nextGraphQLAPIRoute };
