// @keystone-next/keystone/dist/createApolloServer-1b8db11a.cjs.prod.js
'use strict';

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var apolloServerMicro = require('apollo-server-micro');
var apolloServerExpress = require('apollo-server-express');
var session_dist_keystone = require('../session/dist/keystone.cjs.prod.js');

const createApolloServerMicro = ({
  graphQLSchema,
  createContext,
  sessionStrategy,
  graphqlConfig,
  connectionPromise
}) => {
  const context = async ({
    req,
    res
  }) => {
    await connectionPromise;
    return createContext({
      sessionContext: sessionStrategy ? await session_dist_keystone.createSessionContext(sessionStrategy, req, res, createContext) : undefined,
      req
    });
  };

  const serverConfig = _createApolloServerConfig({
    graphQLSchema,
    graphqlConfig
  });

  return new apolloServerMicro.ApolloServer(_objectSpread(_objectSpread({}, serverConfig), {}, {
    context
  }));
};
const createApolloServerExpress = ({
  graphQLSchema,
  createContext,
  sessionStrategy,
  graphqlConfig
}) => {
  const context = async ({
    req,
    res
  }) => createContext({
    sessionContext: sessionStrategy ? await session_dist_keystone.createSessionContext(sessionStrategy, req, res, createContext) : undefined,
    req
  });

  const serverConfig = _createApolloServerConfig({
    graphQLSchema,
    graphqlConfig
  });

  return new apolloServerExpress.ApolloServer(_objectSpread(_objectSpread({}, serverConfig), {}, {
    context
  }));
};

const _createApolloServerConfig = ({
  graphQLSchema,
  graphqlConfig
}) => {
  const apolloConfig = graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.apolloConfig;
  return _objectSpread(_objectSpread({
    schema: graphQLSchema,
    debug: graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.debug
  }, apolloConfig), {}, {
    formatError: formatError(graphqlConfig)
  });
};

const formatError = graphqlConfig => {
  return err => {
    var _graphqlConfig$apollo;

    let debug = graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.debug;

    if (debug === undefined) {
      debug =         "production" !== 'production';
    }

    if (!debug && err.extensions) {
      // Strip out any `debug` extensions
      delete err.extensions.debug;
      delete err.extensions.exception;
    }

    if (graphqlConfig !== null && graphqlConfig !== void 0 && (_graphqlConfig$apollo = graphqlConfig.apolloConfig) !== null && _graphqlConfig$apollo !== void 0 && _graphqlConfig$apollo.formatError) {
      return graphqlConfig.apolloConfig.formatError(err);
    } else {
      return err;
    }
  };
};

exports.createApolloServerExpress = createApolloServerExpress;
exports.createApolloServerMicro = createApolloServerMicro;
